consensus/spos/worker.go

```diff
@@ -757,48 +757,46 @@ func (wrk *Worker) checkValidityAndProcessEquivalentMessages(cnsMsg *consensus.M
 	return nil
 }
 
-func (wrk *Worker) processEquivalentMessageUnprotected(cnsMsg *consensus.Message) error {
-	err := wrk.verifyEquivalentMessageSignature(cnsMsg)
-	if err != nil {
-		return err
+func (wrk *Worker) shouldVerifyEquivalentMessages(msgType consensus.MessageType) bool {
+	if !wrk.enableEpochsHandler.IsFlagEnabled(common.EquivalentMessagesFlag) {
+		return false
 	}
 
+	return wrk.consensusService.IsMessageWithFinalInfo(msgType)
+}
+
+func (wrk *Worker) processEquivalentMessageUnprotected(cnsMsg *consensus.Message) error {
 	hdrHash := string(cnsMsg.BlockHeaderHash)
+	equivalentMsgInfo, ok := wrk.equivalentMessages[hdrHash]
+	if !ok {
+		equivalentMsgInfo = &consensus.EquivalentMessageInfo{}
+		wrk.equivalentMessages[hdrHash] = equivalentMsgInfo
+	}
+	equivalentMsgInfo.NumMessages++
 
-	// if a valid equivalent message was seen before, return error to stop further broadcasts
-	numMessages := wrk.equivalentMessages[hdrHash]
-	wrk.equivalentMessages[hdrHash] = numMessages + 1
-	if numMessages > 0 {
+	if equivalentMsgInfo.Validated {
 		return ErrEquivalentMessageAlreadyReceived
 	}
 
+	err := wrk.verifyEquivalentMessageSignature(cnsMsg)
+	if err != nil {
+		return err
+	}
+
+	equivalentMsgInfo.Validated = true
+
 	return nil
 }
 
-func (wrk *Worker) verifyEquivalentMessageSignature(cnsMsg *consensus.Message) error {
+func (wrk *Worker) verifyEquivalentMessageSignature(_ *consensus.Message) error {
 	if check.IfNil(wrk.consensusState.Header) {
 		return ErrNilHeader
 	}
 
 	header := wrk.consensusState.Header.ShallowClone()
 
-	err := header.SetSignature(cnsMsg.Signature)
-	if err != nil {
-		return err
-	}
-
-	err = header.SetPubKeysBitmap(cnsMsg.PubKeysBitmap)
-	if err != nil {
-		return err
-	}
-
-	err = wrk.headerSigVerifier.VerifySignature(header)
-	if err != nil {
-		log.Debug("verifyEquivalentMessageSignature", "error", err.Error())
-		return err
-	}
-
-	return nil
+	// TODO[Sorin]: after flag enabled, VerifySignature on previous hash, with the signature and bitmap from the proof on cnsMsg
```

Comment:
verify on current header with the signature and bitmap from cnsMsg, but anyway would be removed in next PR.